File: backend/app/api/routes/llm.py
from typing import Any, List, Dict, Optional
from fastapi import APIRouter, Depends, HTTPException, status, BackgroundTasks
from fastapi.responses import StreamingResponse
from sqlalchemy.orm import Session
import json
import asyncio

from app.db.base import get_db
from app.models.user import User
from app.services.llm_service import LLMService
from app.services.llm_config import LLMConfigService
from app.services.embedding_config import EmbeddingConfigService
from app.services.reranking_config import RerankingConfigService
from app.schemas.llm import (
    LLMConfigCreate,
    LLMConfigUpdate,
    LLMConfigResponse,
    LLMProviderResponse,
    LLMProviderInfo,
    ModelsResponse
)
from app.schemas.embedding import (
    EmbeddingConfigCreate,
    EmbeddingConfigUpdate,
    EmbeddingConfigResponse
)
from app.schemas.reranking import (
    RerankingConfigCreate,
    RerankingConfigUpdate,
    RerankingConfigResponse
)
from app.utils.deps import get_current_user, get_current_admin_user
from app.core.config import settings

router = APIRouter()

@router.get("/providers", response_model=Dict[str, Any])
async def get_providers(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> Any:
    """
    Get available LLM providers.
    """
    providers = LLMConfigService.get_available_providers(db)
    return providers

@router.get("/providers/{provider_id}/models", response_model=ModelsResponse)
async def get_provider_models(
    provider_id: str,
    api_key: Optional[str] = None,
    base_url: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> Any:
    """
    Get available models for a specific provider.
    Requires API key for providers that need authentication.
    """
    try:
        # Create a temporary LLM service to get models
        llm_service = LLMService(
            db=db,
            provider=provider_id,
            api_key=api_key,
            base_url=base_url
        )
        
        # Get available models
        chat_models, embedding_models = await llm_service.get_available_models()
        
        return {
            "chat_models": chat_models,
            "embedding_models": embedding_models
        }
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Failed to get models: {str(e)}"
        )

@router.post("/chat/{chat_id}")
async def chat(
    chat_id: str,
    message: str,
    use_rag: bool = True,
    provider: Optional[str] = None,
    model: Optional[str] = None,
    system_prompt: Optional[str] = None,  # Global system prompt for all providers
    temperature: float = 0.7,
    max_tokens: Optional[int] = None,
    stream: bool = True,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> Any:
    """
    Send a message to the LLM and get a response.
    Uses the active LLM configuration if no provider/model is specified.
    The system_prompt parameter is global and applies to all LLM providers.
    """
    # Create LLM service with optional overrides
    # If not provided, it will use the active configuration from the database
    llm_service = LLMService(
        db=db,
        provider=provider,
        model=model,
        system_prompt=system_prompt
    )
    
    # Handle streaming response
    if stream:
        return StreamingResponse(
            stream_chat_response(llm_service, chat_id, message, use_rag, temperature, max_tokens),
            media_type="text/event-stream"
        )
    else:
        # Get response from LLM
        response = await llm_service.chat(
            chat_id=chat_id,
            user_message=message,
            use_rag=use_rag,
            temperature=temperature,
            max_tokens=max_tokens,
            stream=False
        )
        
        return response

async def stream_chat_response(
    llm_service: LLMService,
    chat_id: str,
    message: str,
    use_rag: bool,
    temperature: float,
    max_tokens: Optional[int]
):
    """
    Stream chat response as server-sent events.
    """
    try:
        # Get streaming response from LLM
        async for chunk in llm_service.chat(
            chat_id=chat_id,
            user_message=message,
            use_rag=use_rag,
            temperature=temperature,
            max_tokens=max_tokens,
            stream=True
        ):
            # Format as server-sent event
            yield f"data: {json.dumps(chunk)}\n\n"
            
            # Add a small delay to prevent overwhelming the client
            await asyncio.sleep(0.01)
        
        # End of stream
        yield "data: [DONE]\n\n"
    except Exception as e:
        # Send error as event
        error_data = {
            "error": str(e),
            "done": True
        }
        yield f"data: {json.dumps(error_data)}\n\n"
        yield "data: [DONE]\n\n"

@router.post("/embeddings", response_model=List[List[float]])
async def get_embeddings(
    texts: List[str],
    provider: Optional[str] = None,
    model: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> Any:
    """
    Get embeddings for a list of texts.
    """
    # Use default provider if not specified
    provider = provider or settings.DEFAULT_LLM_PROVIDER
    
    # Create LLM service
    llm_service = LLMService(
        db=db,
        provider=provider,
        model=model
    )
    
    # Get embeddings
    embeddings = await llm_service.get_embeddings(texts)
    return embeddings

@router.post("/admin/config", response_model=LLMConfigResponse)
async def create_llm_config(
    config: LLMConfigCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
) -> Any:
    """
    Create a new LLM configuration. Admin only.
    """
    db_config = LLMConfigService.create_config(db, config)
    return db_config

@router.put("/admin/config/{config_id}", response_model=LLMConfigResponse)
async def update_llm_config(
    config_id: str,
    config: LLMConfigUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
) -> Any:
    """
    Update an existing LLM configuration. Admin only.
    """
    db_config = LLMConfigService.update_config(db, config_id, config)
    if not db_config:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Configuration not found"
        )
    return db_config

@router.get("/admin/config", response_model=List[LLMConfigResponse])
async def get_all_llm_configs(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
) -> Any:
    """
    Get all LLM configurations. Admin only.
    """
    configs = LLMConfigService.get_all_configs(db)
    return configs

@router.get("/admin/config/active", response_model=LLMConfigResponse)
async def get_active_llm_config(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
) -> Any:
    """
    Get the active LLM configuration. Admin only.
    """
    config = LLMConfigService.get_active_config(db)
    if not config:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="No active configuration found"
        )
    return config

@router.post("/admin/config/{config_id}/activate", response_model=LLMConfigResponse)
async def activate_llm_config(
    config_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
) -> Any:
    """
    Set an LLM configuration as active. Admin only.
    """
    config = LLMConfigService.set_active_config(db, config_id)
    if not config:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Configuration not found"
        )
    return config

@router.delete("/admin/config/{config_id}", response_model=Dict[str, Any])
async def delete_llm_config(
    config_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
) -> Any:
    """
    Delete an LLM configuration. Admin only.
    Cannot delete the active configuration.
    """
    try:
        success = LLMConfigService.delete_config(db, config_id)
        if not success:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Configuration not found"
            )
        return {"status": "success", "message": "Configuration deleted"}
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
    
# Embedding Configuration Endpoints
# These endpoints are kept in the LLM router to maintain compatibility with the frontend
@router.post("/admin/embedding/config", response_model=EmbeddingConfigResponse)
async def create_embedding_config(
    config: EmbeddingConfigCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
) -> Any:
    """
    Create a new embedding configuration. Admin only.
    """
    db_config = EmbeddingConfigService.create_config(db, config)
    return db_config

@router.put("/admin/embedding/config/{config_id}", response_model=EmbeddingConfigResponse)
async def update_embedding_config(
    config_id: str,
    config: EmbeddingConfigUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
) -> Any:
    """
    Update an existing embedding configuration. Admin only.
    """
    db_config = EmbeddingConfigService.update_config(db, config_id, config)
    if not db_config:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Configuration not found"
        )
    return db_config

@router.get("/admin/embedding/config", response_model=List[EmbeddingConfigResponse])
async def get_all_embedding_configs(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
) -> Any:
    """
    Get all embedding configurations. Admin only.
    """
    configs = EmbeddingConfigService.get_all_configs(db)
    return configs

@router.get("/admin/embedding/config/active", response_model=EmbeddingConfigResponse)
async def get_active_embedding_config(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
) -> Any:
    """
    Get the active embedding configuration. Admin only.
    """
    config = EmbeddingConfigService.get_active_config(db)
    if not config:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="No active configuration found"
        )
    return config

@router.post("/admin/embedding/config/{config_id}/activate", response_model=EmbeddingConfigResponse)
async def activate_embedding_config(
    config_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
) -> Any:
    """
    Set an embedding configuration as active. Admin only.
    """
    config = EmbeddingConfigService.set_active_config(db, config_id)
    if not config:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Configuration not found"
        )
    return config

@router.delete("/admin/embedding/config/{config_id}", response_model=Dict[str, Any])
async def delete_embedding_config(
    config_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
) -> Any:
    """
    Delete an embedding configuration. Admin only.
    Cannot delete the active configuration.
    """
    try:
        success = EmbeddingConfigService.delete_config(db, config_id)
        if not success:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Configuration not found"
            )
        return {"status": "success", "message": "Configuration deleted"}
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )

# Reranking Configuration Endpoints
# These endpoints are added to the LLM router to maintain consistency with embedding endpoints
@router.post("/admin/reranking/config", response_model=RerankingConfigResponse)
async def create_reranking_config(
    config: RerankingConfigCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
) -> Any:
    """
    Create a new reranking configuration. Admin only.
    """
    db_config = RerankingConfigService.create_config(db, config)
    return db_config

@router.put("/admin/reranking/config/{config_id}", response_model=RerankingConfigResponse)
async def update_reranking_config(
    config_id: str,
    config: RerankingConfigUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
) -> Any:
    """
    Update an existing reranking configuration. Admin only.
    """
    db_config = RerankingConfigService.update_config(db, config_id, config)
    if not db_config:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Configuration not found"
        )
    return db_config

@router.get("/admin/reranking/config", response_model=List[RerankingConfigResponse])
async def get_all_reranking_configs(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
) -> Any:
    """
    Get all reranking configurations. Admin only.
    """
    configs = RerankingConfigService.get_all_configs(db)
    return configs

@router.get("/admin/reranking/config/active", response_model=RerankingConfigResponse)
async def get_active_reranking_config(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
) -> Any:
    """
    Get the active reranking configuration. Admin only.
    """
    config = RerankingConfigService.get_active_config(db)
    if not config:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="No active configuration found"
        )
    return config

@router.post("/admin/reranking/config/{config_id}/activate", response_model=RerankingConfigResponse)
async def activate_reranking_config(
    config_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
) -> Any:
    """
    Set a reranking configuration as active. Admin only.
    """
    config = RerankingConfigService.set_active_config(db, config_id)
    if not config:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Configuration not found"
        )
    return config

@router.delete("/admin/reranking/config/{config_id}", response_model=Dict[str, Any])
async def delete_reranking_config(
    config_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
) -> Any:
    """
    Delete a reranking configuration. Admin only.
    Cannot delete the active configuration.
    """
    try:
        success = RerankingConfigService.delete_config(db, config_id)
        if not success:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Configuration not found"
            )
        return {"status": "success", "message": "Configuration deleted"}
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
            )

File: frontend/services/llm.ts
import { LLMConfig, LLMModel, LLMProvider, EmbeddingConfig, RerankingConfig, RerankingProvider } from '@/types'; // Add RerankingProvider
import { get, post, put, del } from './api';

// ======================
// System Prompt Section
// ======================

export const getSystemPrompt = async (): Promise<{ prompt?: string; error?: string }> => {
  const response = await get<LLMConfig>('/llm/admin/config/active');
  if (response.error || !response.data) {
    return { error: response.error || 'No active config found' };
  }
  return { prompt: response.data.system_prompt };
};

export const updateSystemPrompt = async (prompt: string): Promise<{ success?: boolean; error?: string }> => {
  const activeConfig = await get<LLMConfig>('/llm/admin/config/active');
  if (activeConfig.error || !activeConfig.data) {
    return { error: activeConfig.error || 'No active config found' };
  }
  
  const response = await put<LLMConfig>(`/llm/admin/config/${activeConfig.data.id}`, { system_prompt: prompt });
  if (response.error) {
    return { error: response.error };
  }
  return { success: true };
};

// ======================
// RAG Configuration
// ======================

export const getRAGConfig = async (): Promise<{ config?: any; error?: string }> => {
  const response = await get<LLMConfig>('/llm/admin/config/active');
  if (response.error || !response.data) {
    return { error: response.error || 'No active config found' };
  }
  return { config: response.data.config?.rag_config };
};

export const updateRAGConfig = async (config: any): Promise<{ success?: boolean; error?: string }> => {
  const activeConfig = await get<LLMConfig>('/llm/admin/config/active');
  if (activeConfig.error || !activeConfig.data) {
    return { error: activeConfig.error || 'No active config found' };
  }
  
  const response = await put<LLMConfig>(`/llm/admin/config/${activeConfig.data.id}`, { 
    config: {
      ...(activeConfig.data.config || {}),
      rag_config: config
    }
  });
  if (response.error) {
    return { error: response.error };
  }
  return { success: true };
};

// ======================
// Model Selection
// ======================

// Get all available LLM providers
export const getLLMProviders = async (): Promise<{ providers?: LLMProvider[]; error?: string }> => {
  const response = await get<LLMProvider[]>('/llm/providers');
  if (response.error) {
    return { error: response.error };
  }
  return { providers: response.data };
};

// Get available models for a provider (backward compatibility)
export const getProviderModels = async (
  providerId: string,
  apiKey?: string,
  baseUrl?: string
): Promise<{ chatModels?: string[]; embeddingModels?: string[]; error?: string }> => {
  const params = new URLSearchParams();
  if (apiKey) params.append('api_key', apiKey);
  if (baseUrl) params.append('base_url', baseUrl);
  
  const queryString = params.toString() ? `?${params.toString()}` : '';
  const response = await get<{ chat_models: string[]; embedding_models: string[] }>(
    `/llm/providers/${providerId}/models${queryString}`
  );

  if (response.error) {
    return { error: response.error };
  }
  return {
    chatModels: response.data?.chat_models || [],
    embeddingModels: response.data?.embedding_models || []
  };
};

// Get available chat models for a provider
export const getChatModels = async (
  providerId: string,
  apiKey?: string,
  baseUrl?: string
): Promise<{ models?: string[]; error?: string }> => {
  const result = await getProviderModels(providerId, apiKey, baseUrl);
  if (result.error) {
    return { error: result.error };
  }
  return { models: result.chatModels };
};

// Get available embedding models for a provider
export const getEmbeddingModels = async (
  providerId: string,
  apiKey?: string,
  baseUrl?: string
): Promise<{ models?: string[]; error?: string }> => {
  const result = await getProviderModels(providerId, apiKey, baseUrl);
  if (result.error) {
    return { error: result.error };
  }
  return { models: result.embeddingModels };
};

// ======================
// API Keys Storage
// ======================

export const getAPIKeys = async (): Promise<{ keys?: {api_key?: string; base_url?: string}; error?: string }> => {
  const response = await get<LLMConfig>('/llm/admin/config/active');
  if (response.error || !response.data) {
    return { error: response.error || 'No active config found' };
  }
  return {
    keys: {
      api_key: response.data.api_key,
      base_url: response.data.base_url
    }
  };
};

export const updateAPIKeys = async (keys: {
  api_key?: string;
  base_url?: string;
}): Promise<{ success?: boolean; error?: string }> => {
  const activeConfig = await get<LLMConfig>('/llm/admin/config/active');
  if (activeConfig.error || !activeConfig.data) {
    return { error: activeConfig.error || 'No active config found' };
  }
  
  const response = await put<LLMConfig>(`/llm/admin/config/${activeConfig.data.id}`, keys);
  if (response.error) {
    return { error: response.error };
  }
  return { success: true };
};

// ======================
// Configuration Management
// ======================

export const getActiveLLMConfig = async (): Promise<{ config?: LLMConfig; error?: string }> => {
  const response = await get<LLMConfig>('/llm/admin/config/active');
  if (response.error) {
    return { error: response.error };
  }
  return { config: response.data };
};

export const getAllLLMConfigs = async (): Promise<{ configs?: LLMConfig[]; error?: string }> => {
  const response = await get<LLMConfig[]>('/llm/admin/config');
  if (response.error) {
    return { error: response.error };
  }
  return { configs: response.data };
};

export const createLLMConfig = async (config: {
  chat_provider: string;
  embedding_provider: string;
  model: string;
  embedding_model: string;
  system_prompt: string;
  api_key?: string;
  base_url?: string;
  temperature?: number; // Added temperature
  reranked_top_n?: number | null;
  config?: any;
}): Promise<{ config?: LLMConfig; error?: string }> => {
  const response = await post<LLMConfig>('/llm/admin/config', config);
  if (response.error) {
    return { error: response.error };
  }
  return { config: response.data };
};

export const updateLLMConfig = async (
  configId: string,
  config: {
    chat_provider?: string;
    embedding_provider?: string;
    model?: string;
    embedding_model?: string;
    system_prompt?: string;
    api_key?: string;
    base_url?: string;
    is_active?: boolean;
    temperature?: number; // Added temperature
    reranked_top_n?: number | null;
    config?: any;
  }
): Promise<{ config?: LLMConfig; error?: string }> => {
  const response = await put<LLMConfig>(`/llm/admin/config/${configId}`, config);
  if (response.error) {
    return { error: response.error };
  }
  return { config: response.data };
};

export const activateLLMConfig = async (configId: string): Promise<{ config?: LLMConfig; error?: string }> => {
  const response = await post<LLMConfig>(`/llm/admin/config/${configId}/activate`, {});
  if (response.error) {
    return { error: response.error };
  }
  return { config: response.data };
};

export const deleteLLMConfig = async (configId: string): Promise<{ success?: boolean; error?: string }> => {
  const response = await del(`/llm/admin/config/${configId}`);
  if (response.error) {
    return { error: response.error };
  }
  return { success: true };
};

// ======================
// Embedding Configuration
// ======================

export const getActiveEmbeddingConfig = async (): Promise<{ config?: EmbeddingConfig; error?: string }> => {
  const response = await get<EmbeddingConfig>('/llm/admin/embedding/config/active');
  if (response.error) {
    return { error: response.error };
  }
  return { config: response.data };
};

export const getAllEmbeddingConfigs = async (): Promise<{ configs?: EmbeddingConfig[]; error?: string }> => {
  const response = await get<EmbeddingConfig[]>('/llm/admin/embedding/config');
  if (response.error) {
    return { error: response.error };
  }
  return { configs: response.data };
};

export const createEmbeddingConfig = async (config: {
  provider: string;
  model: string;
  api_key?: string;
  base_url?: string;
  config?: any;
}): Promise<{ config?: EmbeddingConfig; error?: string }> => {
  const response = await post<EmbeddingConfig>('/llm/admin/embedding/config', config);
  if (response.error) {
    return { error: response.error };
  }
  return { config: response.data };
};

export const updateEmbeddingConfig = async (
  configId: string,
  config: {
    provider?: string;
    model?: string;
    api_key?: string;
    base_url?: string;
    is_active?: boolean;
    config?: any;
  }
): Promise<{ config?: EmbeddingConfig; error?: string }> => {
  const response = await put<EmbeddingConfig>(`/llm/admin/embedding/config/${configId}`, config);
  if (response.error) {
    return { error: response.error };
  }
  return { config: response.data };
};

export const activateEmbeddingConfig = async (configId: string): Promise<{ config?: EmbeddingConfig; error?: string }> => {
  const response = await post<EmbeddingConfig>(`/llm/admin/embedding/config/${configId}/activate`, {});
  if (response.error) {
    return { error: response.error };
  }
  return { config: response.data };
};

export const deleteEmbeddingConfig = async (configId: string): Promise<{ success?: boolean; error?: string }> => {
  const response = await del(`/llm/admin/embedding/config/${configId}`);
  if (response.error) {
    return { error: response.error };
  }
  return { success: true };
};

// ======================
// Reranking Configuration
// ======================

export const getActiveRerankingConfig = async (): Promise<{ config?: RerankingConfig; error?: string }> => {
  // Try the new endpoint first
  const response = await get<RerankingConfig>('/llm/admin/reranking/config/active');
  if (!response.error) {
    return { config: response.data };
  }
  
  // Fall back to the old endpoint if the new one fails
  const fallbackResponse = await get<RerankingConfig>('/reranking/active');
  if (fallbackResponse.error) {
    return { error: fallbackResponse.error };
  }
  return { config: fallbackResponse.data };
};

export const getAllRerankingConfigs = async (): Promise<{ configs?: RerankingConfig[]; error?: string }> => {
  // Try the new endpoint first
  const response = await get<RerankingConfig[]>('/llm/admin/reranking/config');
  if (!response.error) {
    return { configs: response.data };
  }
  
  // Fall back to the old endpoint if the new one fails
  const fallbackResponse = await get<RerankingConfig[]>('/reranking');
  if (fallbackResponse.error) {
    return { error: fallbackResponse.error };
  }
  return { configs: fallbackResponse.data };
};

export const createRerankingConfig = async (config: {
  provider: string;
  model: string;
  api_key?: string;
  base_url?: string;
  config?: any;
}): Promise<{ config?: RerankingConfig; error?: string }> => {
  // Try the new endpoint first
  const response = await post<RerankingConfig>('/llm/admin/reranking/config', config);
  if (!response.error) {
    return { config: response.data };
  }
  
  // Fall back to the old endpoint if the new one fails
  const fallbackResponse = await post<RerankingConfig>('/reranking', config);
  if (fallbackResponse.error) {
    return { error: fallbackResponse.error };
  }
  return { config: fallbackResponse.data };
};

export const updateRerankingConfig = async (
  configId: string,
  config: {
    provider?: string;
    model?: string;
    api_key?: string;
    base_url?: string;
    is_active?: boolean;
    config?: any;
  }
): Promise<{ config?: RerankingConfig; error?: string }> => {
  // Try the new endpoint first
  const response = await put<RerankingConfig>(`/llm/admin/reranking/config/${configId}`, config);
  if (!response.error) {
    return { config: response.data };
  }
  
  // Fall back to the old endpoint if the new one fails
  const fallbackResponse = await put<RerankingConfig>(`/reranking/${configId}`, config);
  if (fallbackResponse.error) {
    return { error: fallbackResponse.error };
  }
  return { config: fallbackResponse.data };
};

export const activateRerankingConfig = async (configId: string): Promise<{ config?: RerankingConfig; error?: string }> => {
  // Try the new endpoint first
  const response = await post<RerankingConfig>(`/llm/admin/reranking/config/${configId}/activate`, {});
  if (!response.error) {
    return { config: response.data };
  }
  
  // Fall back to the old endpoint if the new one fails
  const fallbackResponse = await post<RerankingConfig>(`/reranking/${configId}/activate`, {});
  if (fallbackResponse.error) {
    return { error: fallbackResponse.error };
  }
  return { config: fallbackResponse.data };
};

export const deleteRerankingConfig = async (configId: string): Promise<{ success?: boolean; error?: string }> => {
  // Try the new endpoint first
  const response = await del(`/llm/admin/reranking/config/${configId}`);
  if (!response.error) {
    return { success: true };
  }
  
  // Fall back to the old endpoint if the new one fails
  const fallbackResponse = await del(`/reranking/${configId}`);
  if (fallbackResponse.error) {
    return { error: fallbackResponse.error };
  }
  return { success: true };
};

// Get available reranking providers
export const getRerankingProviders = async (): Promise<{ providers?: RerankingProvider[]; error?: string }> => {
  // Use the new dedicated endpoint
  const response = await get<{ providers: RerankingProvider[] }>('/reranking/providers');
  if (response.error) {
    return { error: response.error };
  }
  return { providers: response.data?.providers || [] };
};